### ISSUE TITLE
Exception types for System.Net.Connections

### ISSUE THREAD

**Issue by scalablecory** [dotnet/runtime#39941]
## Background and Motivation

We need an exception model for System.Net.Connections implementations, so that users can catch and react to common network events such as a connection reset.

Normally a `SocketException` would be used, but an abstracted connection may not be using a `Socket` and so that may not be appropriate. Furthermore it might confuse bugs in the connection's code versus events parsed from the protocol that the connection would like to communicate.

This proposes an exception that implementations *should* throw, when appropriate.

## Proposed API

The API is a reduction of the existing `SocketException` and `SocketError` model.

### Option A

We define two new exception types:
- ConnectionException, with a ConnectionError enum, for exceptions from System.Net.Connections APIs
- NetworkException, with a NetworkError enum, for exceptions from Stream.Read/Write[Async].

(We don't love the name NetworkException, but we don't have a better alternative. This is intended to be thrown from connection-oriented Streams like NetworkStream that would be returned from System.Net.Connections APIs. Note, however, that these exceptions are thrown regardless whether the Stream was created via System.Net.Connections or not -- we don't want different semantics based on how the Stream was created.)

```csharp
namespace System.Net.Connections
{
    public class ConnectionException : IOException
    {
        public ConnectionException(string message, ConnectionError connectionError = ConnectionError.Unknown, Exception innerException = null);
        // serialization constructor as usual

        public ConnectionError ConnectionError { get; }
    }

    public enum ConnectionError
    {
        Unknown,
        AddressInUse,       // SocketError.AddressAlreadyInUse
        InvalidAddress,     // SocketError.AddressFamilyNotSupported, SocketError.AddressNotAvailable, etc
        ConnectionRefused,  // SocketError.ConnectionRefused
        HostNotFound,       // SocketError.HostNotFound, SocketError.HostUnreachable, etc
    }
}

namespace System.IO
{
    public class NetworkException : IOException
    {
        public NetworkException(string message, NetworkError error, Exception innerException = null);
        // serialization constructor as usual

        public NetworkError NetworkError { get; }
    }

    public enum NetworkError
    {
        Aborted,
        Reset,
    }
}

```

### Option B

We define one new exception type: NetworkException, with NetworkError enum. System.Net.Connections APIs are expected to throw this exception, as are Stream.Read/Write[Async] in the case of a connection abort or reset. The various differentiated conditions from Option A are merged into a single NetworkError enum.

(Again, we don't love the name; see comments under Option A.)

```csharp
namespace System.IO
{
    public class NetworkException : IOException
    {
        public NetworkException(string message, NetworkError error, Exception innerException = null);
        // serialization constructor as usual

        public NetworkError NetworkError { get; }
    }

    public enum NetworkError
    {
        Unknown,

        AddressInUse,       // SocketError.AddressAlreadyInUse
        InvalidAddress,     // SocketError.AddressFamilyNotSupported, SocketError.AddressNotAvailable, etc
        ConnectionRefused,  // SocketError.ConnectionRefused
        HostNotFound,       // SocketError.HostNotFound, SocketError.HostUnreachable, etc

        Aborted,
        Reset,
    }
}

```

### Option C

We define no new exception type, but instead add an IOError enum property to IOException. System.Net.Connections APIs are expected to throw IOException with an appropriate IOError value, when applicable. Stream.Read/Write[Async] will continue to throw plain IOException, but set the IOError to the appropriate value in the case of a connection abort or reset.

The set of defined IOError values is currently focused on connection-oriented streams, but could be extended to other types of Streams (e.g. FileStream, MemoryStream) in the future.

```csharp
namespace System.IO
{
    public class IOException : Exception
    {
        public IOException(string message, IOError error, Exception innerException = null);

        public IOError IOError { get; }
    }

    public enum IOError
    {
        Unknown,

        AddressInUse,       // SocketError.AddressAlreadyInUse
        InvalidAddress,     // SocketError.AddressFamilyNotSupported, SocketError.AddressNotAvailable, etc
        ConnectionRefused,  // SocketError.ConnectionRefused
        HostNotFound,       // SocketError.HostNotFound, SocketError.HostUnreachable, etc

        Aborted,
        Reset,
    }
}

```

## Notes

`NetworkStream` currently throws `IOException` and would be updated to throw this.

**Comment by ManickaP** [comment]
Does this mean that custom implementations of the new connection API are expected to catch and encapsulate all existing exception (e.g. from sockets)? Or are we gonna do that for them in some of our code? Or will the mix of new and existing exceptions be flowing outside of the API?

Also linking mentioned existing types:
[`SocketException`](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Primitives/src/System/Net/SocketException.cs)
[`SocketError`](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Primitives/src/System/Net/Sockets/SocketError.cs)

**Comment by scalablecory** [comment]
> Does this mean that custom implementations of the new connection API are expected to catch and encapsulate all existing exception (e.g. from sockets)? Or are we gonna do that for them in some of our code? Or will the mix of new and existing exceptions be flowing outside of the API?

This should be the exception that users catch while using this abstraction if they want to respond to network events.

Exceptions should be wrapped if it makes sense to do so. I would not wrap /all/ errors, but rather only the specific ones like ConnectionReset.

**Comment by geoffkizer** [comment]
I think we need to distinguish here between exceptions from the System.Net.Connections APIs, e.g. ConnectAsync/BindAsync/AcceptAsync, and the Stream APIs. For System.Net.Connections, we can do whatever we want. For the existing Stream APIs, we are more constrained.

**Comment by geoffkizer** [comment]
For System.Net.Connections APIs -- that is, ConnectAsync, ListenAsync, AcceptAsync -- here's what I propose:

```csharp
namespace System.Net.Connections
{
    public class ConnectionException : IOException
    {
        public ConnectionException(string message, ConnectionError connectionError = ConnectionError.Unknown, Exception innerException = null);
        // serialization constructor as usual

        public ConnectionError ConnectionError { get; }
    }

    public enum ConnectionError
    {
        Unknown,
        AddressInUse,       // SocketError.AddressAlreadyInUse
        InvalidAddress,     // SocketError.AddressFamilyNotSupported, SocketError.AddressNotAvailable, etc
        ConnectionRefused,  // SocketError.ConnectionRefused
        HostNotFound,       // SocketError.HostNotFound, SocketError.HostUnreachable, etc
    }
}
```

I've intentionally simplified the enum here to try to make it more generic and less Socket-specific. Feedback welcome.

The inner exception here would typically be a SocketException when using sockets underneath, and something else for a custom connection factory.

**Comment by stephentoub** [comment]
> I think we need to distinguish here between exceptions from the System.Net.Connections APIs

Why?

**Comment by geoffkizer** [comment]
> Why?

Well, we don't strictly need to, but it seems useful to me. We're adding a bunch of new API in System.Net.Connections, it has a bunch of error cases that it wants to define, so we should have an Exception (or Exceptions) that live in System.Net.Connections and represent those error cases.

**Comment by stephentoub** [comment]
What errors would it define that we wouldn't also want in a NetworkError enum? I'm not seeing the benefit of the split. Can you elaborate?

**Comment by geoffkizer** [comment]
The Stream APIs (Read/Write) would produce ConnectionReset and ConnectionAborted. The Connection APIs (Connect/Bind/Accept) would produce everything else. No overlap, two different exception types, two different enums.

We can of course mash them all into one exception/enum. It just seems to me it's clearer to have two, since they are used for different purposes.

**Comment by scalablecory** [comment]
> The Stream APIs (Read/Write) would produce ConnectionReset and ConnectionAborted. The Connection APIs (Connect/Bind/Accept) would produce everything else. No overlap, two different exception types, two different enums.

We can't depend on this. This API is designed to be composed -- to allow `ConnectAsync()` to do more than a simple `Socket.Connect()` -- so we must assume some I/O like TLS/SOCKS handshake, CONNECT request, etc. will happen during the Connect.

**Comment by geoffkizer** [comment]
Sure, but it doesn't seem all that useful to be able to report ConnectionReset or ConnectionAborted during ConnectAsync. For one thing, ConnectionAborted is basically meaningless here; what would it mean, that the client logic intentionally sabotaged the connection attempt? ConnectionReset might be slightly useful, but it only really makes sense when there's a layered connect operation, as with a TLS handshake. From the point of view of the user of (e.g.) ConnectAsync, ConnectionReset doesn't mean anything to them generally; it just means that the connection failed to be established, and everything else is just a detail of the underlying transport.

**Comment by geoffkizer** [comment]
I think we're coming at this with different philosophies. I think it's an anti-goal to allow arbitrary error conditions to be exposed from arbitrary operations. Rather, I'd prefer to say: here are a limited number of failure modes you may be interested in -- like abort/reset from Stream ops, and a handful of failure modes (address in use, host not found, connection refused, etc) from Connection ops. I think these constraints actually make the user's life easier, because they don't need to deal with the complexity  of any error happening in any situation, and having to deal with that. After all, the whole point of defining distiguished error codes here is to allow the user to programmatically make some decision based on this, right? Otherwise, we'd just cough up some error text, maybe toss it in a log, and be done with it.

**Comment by geoffkizer** [comment]
> This API is designed to be composed -- to allow ConnectAsync() to do more than a simple Socket.Connect() -- so we must assume some I/O like TLS/SOCKS handshake, CONNECT request, etc. will happen during the Connect.

Just to expand on my point, relative to this:

Yes, API is designed to be entirely general. Literally anything can happen during ConnectAsync. Pigeons could be sent flying.

The fact that there are an endless number of failure modes from the endless number of underlying transports does not mean there's any value in trying to capture all of these failure modes in a single enum.

In other words, having ConnectAsync return ConnectionReset makes as much sense as having ConnectAsync return PigeonWasShotDownInFlight.

**Comment by JamesNK** [comment]
Will this exception only be thrown out of System.Net.Connections APIs, or will it also be thrown from HttpClient.SendAsync? (somewhere in the nested inner exceptions)

I think the gRPC client *might* be interested in this.

There is the concept of connection health status in gRPC and one of the states is `TRANSIENT_FAILURE`. Exception detail that helps identify the difference between "this failed because of a transient error and can be retried" and "this will never work" is useful.

https://github.com/grpc/grpc/blob/master/doc/connectivity-semantics-and-api.md

**Comment by scalablecory** [comment]
> ConnectionReset might be slightly useful, but it only really makes sense when there's a layered connect operation, as with a TLS handshake.

This was a key design goal of the abstraction, we can't be at all hand-wavey about it. Layered connect operations will happen.

(I think your further replies have demonstrated you aren't being hand-wavey, I just want to make sure the expectation is set for all participating)

> From the point of view of the user of (e.g.) ConnectAsync, ConnectionReset doesn't mean anything to them generally; it just means that the connection failed to be established, and everything else is just a detail of the underlying transport.

The people we are targeting in this issue are those that do care about specific network events. Anyone else will probably just be using a `catch(Exception)` or letting things bubble up to a logger.

It may be useful for a user to know the remote party blew up the connection versus simply refusing it.

**Comment by scalablecory** [comment]
> Will this exception only be thrown out of System.Net.Connections APIs, or will it also be thrown from HttpClient.SendAsync? (somewhere in the nested inner exceptions)

I think it's a given that it'll be somewhere in the nested exceptions, but I'd hesitate to guarantee the structure of inner exceptions.

Though, we did do this recently for timeout exceptions.

**Comment by geoffkizer** [comment]
> It may be useful for a user to know the remote party blew up the connection versus simply refusing it.

Sure, that's what inner exceptions are for. The inner exception can tell you all the details. But I'm struggling to see how a general purpose consumer of this API would want to make some sort of programmatic decision about this, and yet would not want to make a programmatic decision about the dead pigeon.

**Comment by scalablecory** [comment]
@geoffkizer will you add to your proposal the 2nd exception type you had in mind for communicating a reset?

**Comment by geoffkizer** [comment]
Sure. Something like:

```csharp
enum NetworkError
{
    Aborted,
    Reset,
}

class NetworkException : IOException
{
    public NetworkException(string message, NetworkError error, Exception innerException = null);

    public NetworkError NetworkError { get; }
}
```

I don't particularly love the name `NetworkException` here, because it feels too narrow; these exceptions should apply to stuff like in-memory pipes too. I suppose we might call it `StreamException`, but that seems too general... `ConnectedStreamException`? That's kinda wordy... suggestions welcome.

**Comment by halter73** [comment]
`IOException` would be great, but I understand not wanting to add another property at this point to such a widely used Exception type. Maybe `ConnectionException`?

**Comment by geoffkizer** [comment]
Let me try to capture the above conversation as well as some offline discussion.

### Goals

1. Allow ASP.NET to detect and handle certain exceptional cases when using System.Net.Connections, in particular:
- AddressInUse from ConnectionListenerFactory.ListenAsync
- ConnectionReset from Stream.Read/Write[Async]. This indicates that the peer aborted the connection.
- ConnectionAborted from Stream.Read/Write[Async]. This indicates that the connection was aborted locally.

Currently, ASP.NET detects these conditions by looking for specific SocketError codes on SocketException. Since System.Net.Connections is a general-purpose abstraction, we need to define a way to surface these conditions independent from SocketException.

2. Define the expected exceptions from the System.Net.Connections APIs.

Currently, we are often throwing IOException from these APIs (which may be fine), but in some cases we may be throwing domain-specific exceptions (which is not fine). We need to codify what exceptions are expected here.

### Option A

We define two new exception types:
- ConnectionException, with a ConnectionError enum, for exceptions from System.Net.Connections APIs
- NetworkException, with a NetworkError enum, for exceptions from Stream.Read/Write[Async].

(We don't love the name NetworkException, but we don't have a better alternative. This is intended to be thrown from connection-oriented Streams like NetworkStream that would be returned from System.Net.Connections APIs. Note, however, that these exceptions are thrown regardless whether the Stream was created via System.Net.Connections or not -- we don't want different semantics based on how the Stream was created.)

```csharp
namespace System.Net.Connections
{
    public class ConnectionException : IOException
    {
        public ConnectionException(string message, ConnectionError connectionError = ConnectionError.Unknown, Exception innerException = null);
        // serialization constructor as usual

        public ConnectionError ConnectionError { get; }
    }

    public enum ConnectionError
    {
        Unknown,
        AddressInUse,       // SocketError.AddressAlreadyInUse
        InvalidAddress,     // SocketError.AddressFamilyNotSupported, SocketError.AddressNotAvailable, etc
        ConnectionRefused,  // SocketError.ConnectionRefused
        HostNotFound,       // SocketError.HostNotFound, SocketError.HostUnreachable, etc
    }
}

namespace System.IO
{
    public class NetworkException : IOException
    {
        public NetworkException(string message, NetworkError error, Exception innerException = null);
        // serialization constructor as usual

        public NetworkError NetworkError { get; }
    }

    public enum NetworkError
    {
        Aborted,
        Reset,
    }
}

```

### Option B

We define one new exception type: NetworkException, with NetworkError enum. System.Net.Connections APIs are expected to throw this exception, as are Stream.Read/Write[Async] in the case of a connection abort or reset. The various differentiated conditions from Option A are merged into a single NetworkError enum.

(Again, we don't love the name; see comments under Option A.)

```csharp
namespace System.IO
{
    public class NetworkException : IOException
    {
        public NetworkException(string message, NetworkError error, Exception innerException = null);
        // serialization constructor as usual

        public NetworkError NetworkError { get; }
    }

    public enum NetworkError
    {
        Unknown,

        AddressInUse,       // SocketError.AddressAlreadyInUse
        InvalidAddress,     // SocketError.AddressFamilyNotSupported, SocketError.AddressNotAvailable, etc
        ConnectionRefused,  // SocketError.ConnectionRefused
        HostNotFound,       // SocketError.HostNotFound, SocketError.HostUnreachable, etc

        Aborted,
        Reset,
    }
}

```

### Option C

We define no new exception type, but instead add an IOError enum property to IOException. System.Net.Connections APIs are expected to throw IOException with an appropriate IOError value, when applicable. Stream.Read/Write[Async] will continue to throw plain IOException, but set the IOError to the appropriate value in the case of a connection abort or reset.

The set of defined IOError values is currently focused on connection-oriented streams, but could be extended to other types of Streams (e.g. FileStream, MemoryStream) in the future.

```csharp
namespace System.IO
{
    public class IOException : Exception
    {
        public IOException(string message, IOError error, Exception innerException = null);

        public IOError IOError { get; }
    }

    public enum IOError
    {
        Unknown,

        AddressInUse,       // SocketError.AddressAlreadyInUse
        InvalidAddress,     // SocketError.AddressFamilyNotSupported, SocketError.AddressNotAvailable, etc
        ConnectionRefused,  // SocketError.ConnectionRefused
        HostNotFound,       // SocketError.HostNotFound, SocketError.HostUnreachable, etc

        Aborted,
        Reset,
    }
}

```

I think this captures all the discussion so far; comments welcome. If there's general agreement that these are the options worth considering, I'll mark as ready for review and we can have an API review discussion of this.

**Comment by scalablecory** [comment]
looks good to me, lets get it set for review today so we can do API review tomorrow

**Comment by jkotas** [comment]
> IOError

Would this also address https://github.com/dotnet/runtime/issues/926 that is about the more general problem of distinguishing different IO errors, not limited to networking errors?

**Comment by antonfirsov** [comment]
I think mixing `FileSystemError` and `NetworkError` stuff in a single `IOError` enum may lead to confusion, and harder debuggability.

**Comment by geoffkizer** [comment]
> Would this also address #926

Possibly, yeah. Certainly we should have a discussion about that since it's a closely related issue.

As @antonfirsov said, I do worry about mixing all sorts of different error codes into a single exception/enum. But it's not obvious where the line should be drawn here, as shown in the above discussion.

**Comment by geoffkizer** [comment]
@danmosemsft This discussion may interest you, as it's related to #926.

**Comment by danmosemsft** [comment]
It would be good to think about where this is going - I've no doubt they will discuss this in API review as well. Currently, there is not much use of managed error code enums in our exceptions.

1. Are we inevitably ending up with one giant list of codes, as in Win32 or Unix? There will certainly be overlap with file IO codes, and codes relating to security (eg access denied). This is visible in the original exception hierarchy of course, and is an advantage of error codes - IO operations today have no single base class to catch, other than Exception itself. We can't reverse time to solve that, but for new API it might help. On the other hand, the proposed [RegexParseError](https://github.com/dotnet/runtime/issues/38872) really is only applicable to regex.

2. Is the error code pattern a good fit for all API and exception types, or only certain cases? Should the property name be domain specific (`NetworkingError`) or should there be a pattern (perhaps `Error` .. we already use `ErrorCode` and of course `NativeErrorCode`). Perhaps, if we mix networking and IO codes in there, it should be the generic name.

3. Should it be considered a breaking change to change an error code? `ConnectionRefused` seems well defined, but one could imagine wanting to break that into finer grained codes. That would break code doing `catch (NetworkException ex) when (ex.Error == NetworkError.ConnectionRefused)` but it would not break code doing `Trace.Write($"Failed because {ex.Error}")`. Finer grained codes can also make it easier to break unintentionally, eg by changing the order of operations.

It feels like API review folks should have a broader discussion about error codes.

**Comment by terrajobst** [comment]
* We'd generally recommend against either extreme (one exception with lots of error codes or as many exceptions as error states) because it creates a mess.
* Instead, we recommend to have one exception type per mainline scenario that the user is likely to write error handling for. It's OK if more advanced cases involve inspecting the exception's error code or inner exception.
* We can even introduce dedicated exception types for specific error codes later, so long they extend the generic type and have the proper error code. If that's the plan, we should consider exposing a factory so that people can construct an exception given an error code. This probably means the constructors are protected.
* It seems *abort* and *reset* cases should have dedicated types.

**Comment by danmosemsft** [comment]
I couldn't make the API review - I'll try to watch the recording, but was #926 considered when rejecting C?

Because if we do #926, we'll end up with an IOException with IOError and a derived NetworkException with both IOError and NetworkError each using their own enums. Is that what we want?

**Comment by geoffkizer** [comment]
The general consensus seemed to be that we didn't want to pollute IOException with a bunch of network-specific error codes.

I assume the same logic would apply to file system errors -- that is, we'd define something like FileSystemException that derives from IOException and has a FileSystemError enum on it.

**Comment by danmosemsft** [comment]
> we'd define something like FileSystemException that derives from IOException and has a FileSystemError enum on it.

Unfortunately IO operations can throw various types derived from IOException, Eg., DirectoryNotFoundException, PathTooLongException. Those would not be FileSystemExceptions, and callers would need to continue to catch them separately in order to distinguish them. If the error codes were on IOException, they would not.

**Comment by scalablecory** [comment]
@danmosemsft from review:
- We should avoid an `IOError` enum that would become a dumping ground for disparate errors.
- Properties on exceptions should generally be informational and not something you would test to make decisions. We should avoid the need for exception filters and prefer concrete types for those exceptions where we expect users to do more than log/bubble.

**Comment by stephentoub** [comment]
> We should avoid the need for exception filters and prefer concrete types for those exceptions where we expect users to do more than log/bubble.

Why? The moment you want to do the same thing for multiple values, it's simpler and cleaner to use a filter.

**Comment by geoffkizer** [comment]
Implemented in #40344
